### DISCUSSION TITLE
Namespacing and subdomain support

### DISCUSSION THREAD

**Discussion by watzon** [luckyframework/lucky#834] [General]
It would be nice to eventually be able to tell certain actions that they are within a namespace, and set everything attached to that namespace to a specific route or subdomain.

For instance, say I have a site that contains the action namespaces `home`, `blog`, and `dashboard`. Currently you can pretty easily designate them to the route prefixes `/`, `/blog`, and `/dashboard`, but it would be nice to be able to instead have `blog` and `dashboard` forward to the subdmomains of the same name.

With rails this can be done with constraints

```ruby
namespace :blog do
  constraints subdomain: 'blog' do
    resources :posts
    resources :tags
    resources :categories
  end
end
```

As far as how to do this best in Lucky, I don't know. Any ideas?

**Comment by paulcsmith** [comment]
This is a great suggestion and I think we should definitely make it an option. Not sure when we can get to it, but it'd be very cool.

I'm imagining something like this:

```crystal
class Blog::Posts::Index < BrowserAction
  route_constraints subdomain: "blog"

  get "/posts" do
  end
end
```

Or to share this between actions, create a `BlogAction`

```crystal
abstract class BlogAction < BrowserAction
  route_constraints subdomain: "blog"
end

class Blog::Posts::Index < BlogAction
  # Do your thing
end
```

I think this API would be nice. It would require some changes to the route handler so that it uses a different router for different domains or something, but we can figure out *how* it works later.

Thoughts @watzon, @jwoertink?

EDIT: We can also add stuff to route constraints to do something like Rails where you can give it a class and do all kinds of fancy stuff...but we can figure that out separately

**Reply by watzon** [reply]
That would be prefect!

**Reply by jwoertink** [reply]
Yeah, I like that too. I wonder if this could be as easy as just using a before pipe and checking the request subdomain 🤔

**Reply by paulcsmith** [reply]
That’s a good point. I think that could work as an mvp for most cases. It could break down though if for example you have two identical paths and one should be handled with a certain sub domain and the other with no subdomain. If you try to do it in a pipe the router won’t look for the other route, it’ll just 404. 

But I think in the meantime that is a good solution and would work today (which is always good!)

**Comment by wout** [comment]
That would be amazing! At some point in the next few months, we will need this. Our api will need to be on a dedicated subdomain for example. So I would be happy to help out.

**Comment by paulcsmith** [comment]
Oh and regarding the route name spacing, I think we should add a “route_prefix” macro that will prefix the routes with whatever you give it. That’ll help with /admin and /api/v1 type namespaces

**Reply by jwoertink** [reply]
this would be great. I currently have to redefine them to make it easier.

**Reply by paulcsmith** [reply]
Yeah I'll add this as an issue since I think that can be separate from the route constraints (and way easier to implement!)

https://github.com/luckyframework/lucky/issues/1087
